### DISCUSSION TITLE
Pull Request Restriction Question

### DISCUSSION THREAD

**Discussion by mborish** [community/community#23306] [New to GitHub]
<p>Hello all,</p>
<p>I’m not familiar with all the ins/outs of Git, and couldn’t find an exact answer to my question.  I am an admin of an organization with numerous repos.  For the sake of this example, we’ll call the repo in question RepoA and say that out of all the admins, I’m primarily responsible for this repo.  On rare occasion, for releases, we merge dev into master.  Now, the issue I have is a user is regularly pushing changes to dev that impact class design in a negative way.  Since everyone is basically working in dev, this causes issues for other users.  As our group grows, I’m sure just letting everyone commit to dev and merge to master won’t be sustainable.</p>
<p>So, based on what I’ve read, the appropriate workflow change would be to make dev/master protected branches and turn on pull requests? (Or at least, this is one option.)  This way, everyone can work in their own branch, merge to dev upon pull request, and I as admin, can merge dev into master for releases.  I can also set the request to require only one review. This way, I have a chance to review changes the user submits to dev before we have to go back and alter code.  Now, the real question I have is, how do I restrict who provides the review? I can turn on “Require pull request reviews before merging” on the settings page, but I don’t see anything that let’s me specify a user?  I’d like to be the sole source of review approval and that includes other admins since I am primarily responsible for this repo.  Do I need to create a CodeOwners file and configure some type of settings?  Or is it even possible to prevent other admins from approving pull requests?  Any suggestion as to how I can setup my workflow to better deal with an overzealous user would be appreciated.  Thanks!</p>

**Comment by guntrip** [comment, marked as answer]
<p>Hello there!</p>
<blockquote>
<p>So, based on what I’ve read, the appropriate workflow change would be to make dev/master protected branches and turn on pull requests? (Or at least, this is one option.)  This way, everyone can work in their own branch, merge to dev upon pull request, and I as admin, can merge dev into master for releases.</p>
</blockquote>
<p>That’s definitely what I’d suggest and would allow you or your colleagues to review any changes before they’re merged into your dev branch.</p>
<blockquote>
<p>Now, the real question I have is, how do I restrict who provides the review? I can turn on “Require pull request reviews before merging” on the settings page, but I don’t see anything that let’s me specify a user?  I’d like to be the sole source of review approval and that includes other admins since I am primarily responsible for this repo.  Do I need to create a CodeOwners file and configure some type of settings?</p>
</blockquote>
<p>With a CODEOWNERS file and protected branches, you can require a review from yourself (or other specified users and teams) before it’s possible to merge a pull request that affects the files and paths specified.</p>
<p>You can get started by setting up a CODEOWNERS file in your repository to define the particular files or paths and users and teams who will own them. You can also use a wildcard to set <em>all files</em> within the repository. The documentation I’m linking below explains the syntax and how to create the file:</p>
<p><a href="https://help.github.com/en/github/creating-cloning-and-archiving-repositories/about-code-owners" rel="nofollow noopener">https://help.github.com/en/github/creating-cloning-and-archiving-repositories/about-code-owners</a></p>
<p>Once this is set up, those users will be automatically requested to review any pull requests that affect the specified files and paths. By enabling “Require reviews” and “Require review from Code Owners” in your repository’s protected branch settings, you can block the ability to merge a branch into your dev branch until one of the users specified in your CODEOWNERS file has left an approving review on the pull request. The documentation below shows the steps for setting this up:</p>
<p><a href="https://help.github.com/en/github/administering-a-repository/enabling-required-reviews-for-pull-requests" rel="nofollow noopener">https://help.github.com/en/github/administering-a-repository/enabling-required-reviews-for-pull-requests</a></p>
<p>I hope this is helpful!</p>

**Comment by mborish** [comment]
<p>Hey Steve,</p>
<p>thanks for the reply this helps.  One other question.  I’ve setup a test repo based on these settings.  However, the issue I run into is that I’m not allowed to approve my own pull requests.  Even on the actual repo, I’m the only admin of note, so merging pull requests into dev/master would fall to me.  Looking around the forums, is there still no way for someone to approve their own pull requests?  If not, wouldn’t the use of a CODEOWNERS file basically require at least two individuals?  I suppose I can always create an alt-account to approve my pulls.  Thanks again! Michael.</p>

**Comment by guntrip** [comment]
<p>Hi again!</p>
<p>Can you try unchecking “Include administrators” in your protected branch settings? This should allow you (and any one else with <strong>admin</strong> permissions) to continue to merge pull requests, including your own, regardless of reviews.</p>

**Comment by mborish** [comment]
<p>Steve,</p>
<p>that works, thanks!  I was confused because regardless of that setting choice.  If you go to review changes, the option to approve your own is still grayed out.  However, the merge on the main pull request page does allow me to do so.</p>
<p>One last point of clarification.  As I mentioned, there are multiple admins, so technically any admin can simply merge with these settings.  If I want to restrict other admins, I could turn that option back on and I have to go with the CODEOWNERS file approach?  Cheers!</p>
